Implementation of Python RTC Test client for Test Harness

raveendra-karu opened the pull request:
This patch includes changes for:
- Python bindings for  webrtc native APIs of libdatachannel
- WebRTC client changes in python chip controller

#### Testing
Step 1:  Build python chip-tool (scripts/build_python.sh)
Step 2:  Build camera-example-app (linux)
Step 3:  Run chip-repl  and verify the webrtc  APIs

#### Result
This PR has been tested manually

raveendra-karu committed:
Implement WebRTC client in Chip controller.

Authors: Raveendra Karu (r.karu@samsung.com)
         Chayan Ailawalia (c.ailawalia@samsung.com)

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Implement Python bindings for WebRTC client.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Update LICENSE and format code.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
remove vscode files.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

yufengwangca commented:
Per offline discussion, Raveendra is updating the WebRTC client to libdatachannel

raveendra-karu committed:
Modify python webrtc bindings to use libdatachannel.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by gn

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Build python webrtc bindings only for linux

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by gn

raveendra-karu committed:
Build python rtc client for only x64

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by gn

raveendra-karu committed:
Add Python test scripts for Provider cluster basic test plans.

Contributors: c.ailawalia@samsung.com, r.karu@samsung.com

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

yufengwangca commented on src/controller/webrtc/BUILD.gn:
```diff
@@ -0,0 +1,50 @@
+
+    deps = [ "${chip_root}/third_party/libdatachannel:libdatachannel" ]
+
+    libs = [
```
We don't use shared lib any more since it will cause runtime error in CI, and I think you should be able to remove this portion since the following line should bring in all needed pieces. 

deps = [ "${chip_root}/third_party/libdatachannel:libdatachannel" ]

raveendra-karu replied:
updated in latest code, pushing in the next commit

yufengwangca commented on src/controller/webrtc/libwebrtc.h:
```diff
@@ -0,0 +1,199 @@
+ * @return void *
+ *     A pointer to the newly created WebRTC client.
+ */
+void * create_webrtc_client(int id);
```
Matter SDK does not use this function name style, I saw you already use the correct style such as "CreateOffer", please algin the rest of function and use consistent naming style

raveendra-karu replied:
Done

yufengwangca commented on src/controller/webrtc/libwebrtc.cpp:
```diff
@@ -0,0 +1,290 @@
+    // Set up event handlers for the peer connection
+    client->pc->onLocalDescription([client](rtc::Description description) {
+        std::cout << "Local Description (Paste this to the other peer):" << std::endl;
+        std::cout << std::string(description) << std::endl;
```
We do not use std::cout within the SDK, please update with ChipLogProgress

raveendra-karu replied:
done

yufengwangca commented on src/controller/webrtc/libwebrtc.h:
```diff
@@ -0,0 +1,199 @@
+
+#include "callbacks.h"
+
+namespace chipwebrtc {
```
The namespace should be following to align with SDK 
```
namespace chip {
namespace WebRTC {
```

raveendra-karu replied:
done

yufengwangca commented on src/controller/webrtc/BUILD.gn:
```diff
@@ -0,0 +1,50 @@
+import("${chip_root}/build/chip/buildconfig_header.gni")
+
+if (current_os == "linux" && current_cpu == "x64") {
+  static_library("chip_webrtc") {
```
We should not use if to declare the static_library, but reference the static_library. Remove this check.

raveendra-karu replied:
done

yufengwangca commented on src/controller/webrtc/BUILD.gn:
```diff
@@ -0,0 +1,50 @@
+if (current_os == "linux" && current_cpu == "x64") {
+  static_library("chip_webrtc") {
+    sources = [
+      "libwebrtc.cpp",
```
Change the file name to Matter SDK style, Callbacks.h, WebRTC.app, WebRTC.h

raveendra-karu replied:
done

yufengwangca commented on src/controller/webrtc/libwebrtc.cpp:
```diff
@@ -0,0 +1,290 @@
+
+namespace chipwebrtc {
+
+class RTCClient
```
Rename it to WebRTCClient to differentiate with gRPC used in Matter

raveendra-karu replied:
done

yufengwangca commented on src/controller/webrtc/libwebrtc.cpp:
```diff
@@ -0,0 +1,290 @@
+    int client_id;
+    std::shared_ptr<rtc::PeerConnection> pc;
+    rtc::Configuration config;
+    SdpOfferCallback offer_cb;
```
Matter use camel case to name member variables, such as 'systemState'

raveendra-karu replied:
done

raveendra-karu replied:
done

raveendra-karu committed:
Address review comments.

Authors: Chayan Ailawalia <c.ailawalia@samsung.com>
         Raveendra Karu <r.karu@samsung.com>

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

yufengwangca commented on src/controller/webrtc/BUILD.gn:
```diff
@@ -0,0 +1,49 @@
+  deps = [ "${chip_root}/third_party/libdatachannel:libdatachannel", "${chip_root}/src/lib/support", ]
+
+  lib_dirs = [ rebase_path("${chip_root}/third_party/libdatachannel/repo/build/") ]
+
```
Those are not needed, and we don't build shared lib any more, you should not be able to link it

raveendra-karu replied:
done

raveendra-karu replied:
done

yufengwangca commented on src/controller/webrtc/callbacks.h:
```diff
@@ -0,0 +1,26 @@
+/*
```
this file needs to be deleted

raveendra-karu replied:
okay

arkq replied:
You've resolved the comment, but the file is still here....

raveendra-karu replied:
yes, missed it earlier. now removed.

yufengwangca commented on src/controller/webrtc/libwebrtc.h:
```diff
@@ -0,0 +1,199 @@
+/*
+ *
+ *    Copyright (c) 2025 Project CHIP Authors
```
This file need to be deleted

raveendra-karu replied:
okay

raveendra-karu committed:
Resolve ssl, crypto dependency errors.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu commented on src/controller/webrtc/libwebrtc.cpp:
```diff
@@ -0,0 +1,290 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
```
removed

restyled-commits committed:
Restyled by gn

arkq commented on src/controller/python/chip/webrtc/callbacks.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def on_ice(candidates, peer):
+    print("on_ice called")
```
Use `logging` instead of `print` in all places.

raveendra-karu replied:
sure, done.

arkq commented on src/controller/python/chip/webrtc/types.py:
```diff
@@ -0,0 +1,62 @@
+#  limitations under the License.
+#
+
+from ctypes import *
```
Import all used types, so static analyzers can verify imports.

raveendra-karu replied:
Done

arkq commented on src/controller/webrtc/BUILD.gn:
```diff
@@ -0,0 +1,51 @@
+    "WebRTC.h",
+  ]
+  cflags_cc = [
+    "-Wno-shadow",
```
Fix issues in the code instead of disabling warnings.

raveendra-karu replied:
If I enable the flags, build errors occurs in the third_party (libdatachannel) code . Hence using the flags.
ex: Build error
```
../../third_party/libdatachannel/repo/include/rtc/frameinfo.hpp: In constructor ‘rtc::FrameInfo::FrameInfo(uint32_t)’:
../../third_party/libdatachannel/repo/include/rtc/frameinfo.hpp:19:28: error: declaration of ‘timestamp’ shadows a member of ‘rtc::FrameInfo’ [-Werror=shadow]
   19 |         FrameInfo(uint32_t timestamp) : timestamp(timestamp) {};
      |                   ~~~~~~~~
```

arkq replied:
If the problem is with `libdatachannel` all these flags should be only for `libdatachannel` not for our code. You will have to fix the `config("datachannel_config")` in `third_party/libdatachannel/BUILD.gn` to properly define public configuration for that library. If any warnings need to be excluded, you have to do it there, not here.

Also, you should not add `/third_party/libdatachannel/repo/include` to the `include_dirs` manually. This should be provided by the `/third_party/libdatachannel:libdatachannel` dependency configuration.

arkq commented on src/controller/webrtc/WebRTC.cpp:
```diff
@@ -0,0 +1,300 @@
+    FILE * file = fopen(filename, "r");
+    if (!file)
+    {
+        printf("Could not open %s\n", filename);
```
Use proper logging.

raveendra-karu replied:
Removed this code now

arkq commented on src/controller/webrtc/WebRTC.cpp:
```diff
@@ -0,0 +1,300 @@
+    WebRTCClient(int id) { client_id = id; }
+};
+
+// Function to read file content into a string
```
Add to the docstring that the string needs to be freed with `free`. However, better yet, use our `MemoryAlloc` from `lib/suport/CHIPMem.h`

raveendra-karu replied:
readFile() API is added for test purposes. removed now.

arkq commented on src/controller/webrtc/WebRTC.cpp:
```diff
@@ -0,0 +1,300 @@
+    WebRTCClient * client = static_cast<WebRTCClient *>(Client);
+
+    // Add default ICE servers to the configuration
+    client->config.iceServers.emplace_back("stun.l.google.com:19302");
```
Why hardcoded host?

raveendra-karu replied:
For verifying public ice, removed now.

arkq reviewed (changes requested):
I think that it would be better to create a proper WebRTC class instead of C-like API (with standalone functions in the `chip::webrtc` namespace).

raveendra-karu committed:
Address review comments.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu commented:
> I think that it would be better to create a proper WebRTC class instead of C-like API (with standalone functions in the `chip::webrtc` namespace).

Thanks for your suggestion. I'll plan that in subsequent PRs.
Currently, some Python TCs are implemented with CAPI interface. Modification of APIs leads to update all TCs.

raveendra-karu committed:
Address review comments.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Fix Lint issues.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

arkq commented on src/controller/webrtc/WebRTC.cpp:
```diff
@@ -0,0 +1,270 @@
+    }
+
+    rtcClient->pc->close();
+    delete rtcClient;
```
Having dedicated destroy function which does nothing and a close function which actually destroys the object is kinda misleading...

raveendra-karu committed:
Address review comments.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Fix build errors.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Install cmake using pip.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Fix LINT issues and CI build failures.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Address review comments suggested by copilot.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Refactor Python webrtc bindings.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

raveendra-karu committed:
Address review comments.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Fix LINT errors.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

bzbarsky-apple commented on src/app/clusters/closure-control-server/closure-control-cluster-logic.h:
This is not related to this PR as described.  Please remove all the irrelevant changes.

restyled-commits committed:
Restyled by autopep8

raveendra-karu commented on scripts/build_python.sh:
```diff
@@ -49,7 +49,7 @@ declare install_pytest_requirements=yes
 declare -a extra_packages
 declare -a extra_gn_args
-
+chip_support_webrtc_python_bindings=$chip_support_webrtc_python_bindings
```
Done

raveendra-karu commented on src/controller/webrtc/WebRTC.h:
```diff
@@ -0,0 +1,200 @@
+ * @return void *
+ *     A pointer to the newly created WebRTC client.
+ */
+void * CreateWebrtcClient(int id);
```
Done the changes and updated PR

raveendra-karu commented on src/controller/webrtc/WebRTC.cpp:
```diff
@@ -0,0 +1,273 @@
+}
+
+// Function to initialize the peer connection
+void InitialisePeerConnection(void * Client)
```
Refactored the changes with c++ implementation and CAPI for the same. Could you please review

raveendra-karu commented on src/controller/python/chip/webrtc/__init__.py:
```diff
@@ -0,0 +1,80 @@
+from chip.webrtc.library_handle import _GetWebRTCLibraryHandle
+
+
+def CreateWebrtcClient(client_id):
```
Done

raveendra-karu commented on scripts/build_python.sh:
```diff
@@ -49,7 +49,7 @@ declare install_pytest_requirements=yes
 declare -a extra_packages
 declare -a extra_gn_args
-
+chip_support_webrtc_python_bindings=$chip_support_webrtc_python_bindings
```
This is to use flag value set in controller/flags.ini and define gn args accordingly. Updated in the latest PR.

raveendra-karu committed:
Fix LINT errors.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu commented:
Hi @bzbarsky-apple, accidentally other changes were included during earlier rebase.
As you suggested, removed non-related changes from this PR. Please take a look!

arkq commented on scripts/build_python.sh:
```diff
@@ -64,6 +64,10 @@ case "$OS_TYPE" in
 esac
 
+if [ -n "$chip_support_webrtc_python_bindings" ]; then
+    extra_gn_args+=("--args=chip_support_webrtc_python_bindings=true")
```
This will not work. Wait for the changes in here https://github.com/project-chip/connectedhomeip/pull/39578 and add:
```bash
gn_args+=("chip_support_webrtc_python_bindings=true")
```

arkq commented on src/controller/BUILD.gn:
```diff
@@ -118,6 +118,10 @@ static_library("controller") {
 
   cflags = [ "-Wconversion" ]
+  if (chip_support_webrtc_python_bindings && current_os == "linux" &&
+      (current_cpu == "x64" || current_cpu == "armv64")) {
```
There is no `armv64` from where did you get that value? Also, again, why the CPU constraint here?

raveendra-karu replied:
It's a typo! 
Corrected it as "arm64"!

As of now, Camera app is built and verified on Linux and ARM.
Only camera app uses "libdatachannel & webrtc". Hence limiting bindings only to these 2 architectures.

arkq replied:
Sorry, but limiting per architecture does not make any sense here. Do you need some CPU feature which is not available on let say RISC-V?

arkq commented on src/controller/python/BUILD.gn:
```diff
@@ -93,6 +93,10 @@ shared_library("ChipDeviceCtrl") {
       "chip/utils/DeviceProxyUtils.cpp",
     ]
+    if (chip_support_webrtc_python_bindings && current_os == "linux" &&
+        (current_cpu == "x64" || current_cpu == "armv64")) {
```
Why CPU constraint?

arkq replied:
> As of now, Camera app is built and verified on Linux and ARM.

But are there any technical reasons why we want this constraint?

raveendra-karu replied:
Yes. `libdatachannel` is not built for other platforms. this is the dependency.

arkq commented on .github/workflows/lint.yml:
```diff
@@ -122,6 +122,7 @@ jobs:
                      --known-failure app/util/util.cpp \
                      --known-failure app/util/util.h \
                      --known-failure app/WriteHandler.h \
+                     --known-failure controller/webrtc/Callbacks.h \
```
This list is only for file that are already in the SDK. Do not add new files which are not part of any BUILD.gn

arkq commented on src/controller/webrtc/WebRTCClient.h:
```diff
@@ -0,0 +1,51 @@
+    void onIceCandidate(std::function<void(const std::string &, const std::string &)> callback);
+
+private:
+    std::shared_ptr<rtc::PeerConnection> pc_;
```
Please follow the convention in the codebase:
- members: `mCammelCase`
- methods `CammelCase`

arkq replied:
Why `shared_ptr` ? I can not see any place that this connection instance is shared anywhere...

arkq commented on src/controller/webrtc/WebRTCClient.cpp:
```diff
@@ -0,0 +1,86 @@
+    {
+        ChipLogError(NotSpecified, "No STUN server URL provided");
+    }
+    pc_ = std::make_shared<rtc::PeerConnection>(config);
```
Why connection is not created in the client constructor? What will happen if someone will call `createPeerConnection` on the same client instance?

raveendra-karu replied:
Generally, peerconnection creation is decoupled from client creation.
This will be useful in scenarios such as,
-  clients can do some initial signaling to receive ice servers from ecosystem
-  client need to wait till sdp offer is received, then creates peer connection and then invokes create answer when acts as receive


Handled case of multiple times invocation of `CreatePeerConnection` API

arkq commented on src/controller/webrtc/WebRTC.cpp:
```diff
@@ -0,0 +1,127 @@
+static std::mutex g_mutex;
+static std::map<WebRTCClientHandle, std::shared_ptr<WebRTCClient>> g_clients;
+
+WebRTCClientHandle webrtc_client_create()
```
Why this is not a proper C++ class? Matter SDK is a C++ project, do not create C API if not required.

arkq replied:
Yes, but this is not Python binding, this is still in the matter SDK namespace. You have python binding in `src/controller/python/chip/webrtc/RTCClient.cpp`. Using C in `python/chip/webrtc/RTCClient.cpp` is perfectly fine fore me. However, this C binding should use C++ from matter SDK.

raveendra-karu committed:
Address review comments.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Removed Callbacks.h as it is unused.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

arkq commented on src/controller/webrtc/WebRTCClient.cpp:
```diff
@@ -0,0 +1,126 @@
+    if (mPeerConnection == nullptr)
+    {
+        ChipLogError(NotSpecified, "Failed to create PeerConnection");
+        return;
```
If you don't want to integrate this method into the constructor, at least return `CHIP_ERROR` in case of failure, so the user will now that connection creation has failed.

raveendra-karu replied:
Done

raveendra-karu commented on src/controller/webrtc/WebRTC.h:
```diff
@@ -0,0 +1,124 @@
+ *
+ * @return WebRTCClientHandle Opaque handle to the created client instance.
+ */
+WebRTCClientHandle webrtc_client_create();
```
Yes, only in bindings it will be used.
Tried similar approach of moving this CAPI layer to python native,
As during python package building only APIs declarations of CAPI layer is accessed,
WebRTC native package is build and linked. So If I move this implementation to Python,  then ``WebRTCClient.h` header related build error occurs as python native module couldn't find this header file.

I would consider these changes in further PRs by checking any alternative approach is there.

arkq replied:
> As during python package building only APIs declarations of CAPI layer is accessed,
WebRTC native package is build and linked. So If I move this implementation to Python, then ``WebRTCClient.h` header related build error occurs as python native module couldn't find this header file.

I've just verified moving WebRTC.cpp into python binding and everything compiled and linked properly. If you are not going to use this WebRTC.h anywhere except Python binding, please, move all the logic there.

raveendra-karu committed:
Return CHIP_ERROR for peer connnection creation failure

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by isort

arkq commented on src/controller/flags.gni:
```diff
@@ -17,4 +17,7 @@ import("//build_overrides/chip.gni")
   chip_support_commissioning_in_controller = true
+
+  # Build controller with webrtc python bindings
+  chip_support_webrtc_python_bindings = false
```
Please make sure that this is set to true in CI, otherwise Python webrtc support will not be check by CI.

arkq replied:
You should include that now. Otherwise, you will have to state in the `Testing` section in PR description that this PR was tested manually. However, current statement is incorrect - running `scripts/build_python.sh` will not compile your changes...

Every PR needs to be tested, and if it is possible, we require automatic testing by CI. Adding, `-webrtc` option to the `build_examples.py` is simple enough that you should include it in that PR.

Also, please do not resolve conversation if it was not in fact resolved... This makes tracking review process really hard...

enkiusz commented on src/controller/webrtc/WebRTC.cpp:
```diff
@@ -0,0 +1,126 @@
+namespace webrtc {
+
+static std::mutex g_mutex;
+static std::map<WebRTCClientHandle, std::shared_ptr<WebRTCClient>> g_clients;
```
Why do we need the map and shared_ptr here?
Why not just create WebRTCClient here and return it ? See `src/controller/python/chip/crypto/p256keypair.cpp` for a similar binding for managing C++ object lifecycle.

raveendra-karu committed:
Address review comments of logger

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Remove C APIs from Matter SDK code

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

raveendra-karu committed:
Revert the unwanted changes.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Removed not required flags

raveendra-karu committed:
Fix LINT error.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by isort

raveendra-karu committed:
Fix LINT error in library handle.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by isort

raveendra-karu committed:
Fix LINT errors.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by isort

enkiusz reviewed (changes requested):
pychip_webrtc_client_set_local_description_callback and pychip_webrtc_client_set_ice_candidate_callback are still not wired properly in `src/controller/python/chip/webrtc/__init__.py`

raveendra-karu committed:
Fix LINT error.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Add missing callback functions.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Fix Typo.

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by autopep8

raveendra-karu committed:
Add python webrtc building build capabilities to build_examples.py

* Add python webrc bindings build support into build_examples.py
* Ensure CI will also build python webrtc bindings

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

arkq commented on scripts/build/build/targets.py:
```diff
@@ -127,6 +127,7 @@ def BuildHostTarget():
         TargetPart('simulated-app2', app=HostApp.SIMULATED_APP2,
                    enable_ble=False, enable_wifipaf=False),
         TargetPart('python-bindings', app=HostApp.PYTHON_BINDINGS),
+        TargetPart('python-bindings-webrtc', app=HostApp.PYTHON_BINDINGS_WEBRTC),
```
You need to add ` target.AppendModifier('webrtc', enable_webrtc=True)` not `TargetPart`

raveendra-karu committed:
append webrtc to target.AppendModifier

raveendra-karu committed:
Fix build errors

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Fixed build error

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

arkq commented on scripts/build/builders/host.py:
```diff
@@ -506,6 +507,9 @@ def __init__(self, root, runner, app: HostApp, board=HostBoard.NATIVE,
         if chip_casting_simplified is not None:
             self.extra_gn_options.append(f'chip_casting_simplified={str(chip_casting_simplified).lower()}')
 
+        if enable_webrtc is not None:
```
```suggestion
        if enable_webrtc:
```

raveendra-karu committed:
Fix webrtc is not none issue

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

raveendra-karu committed:
Remove |linux-x64-python-bindings-webrtc| from yaml

Signed-off-by: Raveendra Karu <r.karu@samsung.com>

arkq committed:
Apply suggestions from code review
